tasm-lib/src/list/higher_order/map.rs

```diff
@@ -1,330 +1,346 @@
-
-                dup 1
-                dup 1
-                // _ <aia> *input_list len *input_list len
-
-                {&mul_elem_size(input_elem_size)}
-                add
-                // _ <aia> *input_list len *input_list_last_word
-
-                call {new_list}
-                // _ <aia> *input_list len *input_list_last_word *output_list
+        let main_loop_body = triton_asm! {
+            // INVARIANT: _ *end_condition_in_list *output_elem *input_elem
+            {main_loop_fn}:
 
-                // Write output list's length
-                dup 2
-                dup 1
-                write_mem 1
-                pop 1
+            /* maybe return */
+            // not using `recurse_or_return` to have more room for parameters
+            // that might live on the stack, to and used by the inner function
+            dup 2 dup 1 eq
+            skiz return
 
-                dup 2
-                {&mul_elem_size(output_elem_size)}
-                add
-                // _ <aia> *input_list len *input_list_last_word *output_list_last_word_last_element
+            /* read */
+            {&input_type.read_value_from_memory_leave_pointer()}
+            place {input_elem_size}
+                        // _ *end_condition_in_list *output_elem *prev_input_elem [input_elem]
 
-                {&adjust_output_list_pointer}
-                // _ <aia> *input_list len *input_list_last_word *output_list_first_word_last_element
+            /* map */
+            {&exec_or_call_inner_function}
+                        // _ *end_condition_in_list *output_elem *prev_input_elem [output_elem]
 
-                swap 2
-                pop 1
-                // _ <aia> *input_list *output_list_first_word_last_element *input_list_last_word
+            /* write */
+            pick {output_type.stack_size() + 1}
+            {&output_type.write_value_to_memory_leave_pointer()}
+            addi {-2 * output_type.stack_size() as i32}
+            place 1     // _ *end_condition_in_list *prev_output_elem *prev_input_elem
 
-                call {main_loop}
-
-                pop 1
-                {&final_output_list_pointer_adjust}
+            recurse
+        };
 
-                swap 1
-                pop 1
+        let map_one_list = triton_asm! {
+            // BEFORE: _ [fill; M]   [*in_list; N-M]   *out_list
+            // AFTER:  _ [fill; M+1] [*in_list; N-M-1] *out_list
+
+            /* read list lengths */
+            read_mem 1
+            addi 1      // _ [_; M] [_; N-M-1] *in_list out_list_len *out_list
+
+            pick 2
+            read_mem 1
+            addi 1      // _ [_; M] [_; N-M-1] out_list_len *out_list in_list_len *in_list
+
+            /* prepare in_list pointer for main loop */
+            dup 1
+            {&mul_elem_size(input_elem_size)}
+            dup 1
+            add         // _ [_; M] [_; N-M-1] out_list_len *out_list in_list_len *in_list *in_list_last_word
```

Comment:
The name is mostly there because I re-used the existing terminology of the `map` code I generalized. However, this is only an explanation for where the name comes from and not an argument against changing it, since of course any refactoring should be taken as a chance to improve code quality.

`*in_list_first_elem_last_word` is consistent with other terminology in use. :+1: